### PR TITLE
Import articles

### LINKED ISSUE #392: Import useful blog posts as docs pages in "Articles" section
## Why are we implementing it? (sales eng)

There is a lot of great technical writing on our blog, but it can be harder to find than official documentation. Making docs pages out of posts allows them to be searched more easily, and prevents them from being "buried" over time.

## How does this work?

- [x] Review posts to find top 10 technical posts
- [ ] Identify which (potentially new) sections they should go in
- [ ] Remove redundant intro content that describes Citus, leaving the meat of the articles

### PR DESCRIPTION
**Pull request by begriffs**
Fixes #392 

These are the articles I imported

* https://www.citusdata.com/blog/2017/07/13/citus-addon-for-heroku/
  * Turned it into a paragraph in our installation section
* https://www.citusdata.com/blog/2017/01/17/parallel-indexing-with-citus/
  * Added it as a Table Management topic
* https://www.citusdata.com/blog/2016/04/28/scalable-product-search/
  * Turned it into the "Real-Time Faceted Search" tutorial
* https://www.citusdata.com/blog/2017/03/10/how-to-scale-postgresql-on-aws/
  * Used it as the new cloud intro, "Why We Built Cloud"
* https://www.citusdata.com/blog/2016/11/29/event-aggregation-at-scale-with-postgresql/
  * Converted to "Real-Time Event Aggregation" tutorial

I omitted the next batch of articles. They turned out to be just too random, or else more about postgres than Citus:

* https://www.citusdata.com/blog/2016/10/10/outer-joins-in-citus/
* https://www.citusdata.com/blog/2017/06/30/efficient-rollup-with-hyperloglog-on-postgres/
* https://www.citusdata.com/blog/2016/08/30/citus-clouds-usage-of-citus-cloud/
* https://www.citusdata.com/blog/2016/07/25/sharding-json-in-postgres-and-performance/

This next batch was redundant with the information in docs already. I think redundancy is already kind of a problem in docs, so I didn't want to make it worse.

* https://www.citusdata.com/blog/2017/04/04/distributed_count_distinct_with_postgresql/
* https://www.citusdata.com/blog/2016/10/03/designing-your-saas-database-for-high-scalability/
* https://www.citusdata.com/blog/2016/08/10/sharding-for-a-multi-tenant-app-with-postgres/

### CODE REVIEW

**Comment by begriffs** [comment]
Note to self: there are sphinx warnings we'll want to fix, but I'm de-prioritizing this PR while others review which of the blog posts would be better in a separate "articles" section.

**Comment by ozgune** [comment]
Hey @begriffs -- I read over these changes. I think most of these blog posts (and the posts that we didn't yet incorporate) could live better in an Articles section. Here's why I think this way.

1. https://www.citusdata.com/blog/2016/04/28/scalable-product-search: Turned it into the "Real-Time Faceted Search" tutorial
2. https://www.citusdata.com/blog/2016/11/29/event-aggregation-at-scale-with-postgresql: Converted to "Real-Time Event Aggregation" tutorial

In our current tutorials, we provide step-by-step instructions. Our first goal is to guide users to download sample data and make them successful at the end of a tutorial. Our secondary goal is to prime them for our target use cases. Neither of these blog posts fit that pattern -- that said, I think they are pretty interesting for more advanced users.

(Side question: Have we thought about updating the Use-case Guide > Real-time Analytics with information in the Real-time Event Aggregation tutorial?)

3. https://www.citusdata.com/blog/2017/03/10/how-to-scale-postgresql-on-aws: Used it as the new cloud intro, "Why We Built Cloud"

This section talks to Citus Cloud's benefits and our architectural decisions in delivering those benefits. In that sense, a more accurate description could be "Why We Built Citus Cloud In a Particular Way." I don't know where other databases put architectural decisions and their resulting benefits. This feels more product marketing rather than traditional docs material to me. But I could be convinced. I know for example PostgREST does this. :)

4. https://www.citusdata.com/blog/2017/01/17/parallel-indexing-with-citus: Added it as a Table Management topic

If we had a section on benchmarking, this blog post could go there. That said, I think we'd need to do some rewording to align this post's contents with that new section.

5. https://www.citusdata.com/blog/2017/07/13/citus-addon-for-heroku: Turned it into a paragraph in our installation section

I'm split on this post. Part of this post talks about Citus Cloud's benefits and we also talked about them in [the Citus Cloud launch post](https://www.citusdata.com/blog/2016/04/19/introducing-citus-cloud/).

I don't know if we'd like to mention Heroku as a way to install Citus. I think not -- so that we don't conflate this section further.

File: cloud/scaling.rst [hunk 1]
```diff
@@ -1,3 +1,21 @@
...
+.. image:: ../images/cloud_provisioning.png
+
+Configuring Your Plan
+---------------------
+
+Citus Cloud plans vary based on the size of your primary node, size of your distributed nodes, number of distributed nodes and whether you have high availability or not. From within the Citus console you can configure your plan or you can preview what it might look like within the `pricing calculator <https://console.citusdata.com/pricing>`_.
+
+The key items you'll care about for each node:
+
+- Storage - All nodes come with 512 GB of storage
```
Comment (ozgune) [inline cloud/scaling.rst]: I think we increased the storage size to 1 TB.

File: cloud/scaling.rst [hunk 2]
```diff
@@ -1,3 +1,21 @@
...
+- Memory - The memory on each node varies based on the size of node you select
+- Cores - The cores on each node varies based on the size of node you select
```
Comment (ozgune) [inline cloud/scaling.rst]: Physical cpu cores? (We make a deliberate attempt to use physical rather than virtual cores. So I just wanted to ask.)

**Comment by ozgune** [comment]
@begriffs -- I think the fastest way to move forward on this PR is to open a new Articles section and start adding these blog posts into that section. What do you think?

**Comment by begriffs** [comment]
Sure, no problem. I'll move them all into an articles section.

**Comment by begriffs** [comment]
Articles are in their own section now. Ready for review.

**Comment by ozgune** [comment]
Hey @begriffs - I went over this pull request today. This looks generally good.

I had the following comments.

* I wasn't able to see an active deployment / preview. How can we get a preview?
* Since the content in the articles section is the same as the one in our blog, do we need to do anything on the SEO side?
* We shortened or changed the names for these blog posts. Could we keep the original titles? (Similarly, one the of links was too short -- aggregation.rst. Could we make it a bit more descriptive?)
* We skipped over some of the articles we originally picked. For example, we skipped over the sharding for a multi-tenant app and HLL articles. Could we include them?

File: articles/designing_saas.rst [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+The good news is, databases have advanced quite a bit in the past ten
+years in accommodating SaaS applications at scale. What was once only
+available to the likes of Google and Salesforce with significant
+engineering effort, is now becoming accessible to everyone with
+open-source technologies such as PostgreSQL and
+`Citus <https://github.com/citusdata/citus>`__. If you’re thinking of
+modernizing your multi-tenant architecture, `drop us a
+line </about/contact_us>`__ and we’d be happy to chat.
+
+.. raw:: html
```
Comment (ozgune) [inline articles/designing_saas.rst]: I don't think we need the section about signing up for our newsletter in our documentation.

File: articles/heroku_addon.rst [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+the Heroku team know and we'll make sure to get it created and available
+for you.
+
+We want scaling your Postgres database on Heroku to be as easy as
+scaling your Heroku app. Which is why we're excited to roll out the
+Citus add-on to help you grow. If you’re using Heroku Postgres and are
+looking for scale, `give the Citus add-on for
+Heroku <https://elements.heroku.com/addons/citus>`__ a try today.
+
+p.s. And if you have any questions at all for our team at Citus Data,
```
Comment (ozgune) [inline articles/heroku_addon.rst]: Maybe we take out the P.S. part?

File: articles/outer_joins.rst [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+columns <https://docs.citusdata.com/en/v5.2/dist_tables/concepts.html>`__
+of the distributed tables. - The query should join the distributed
+tables on the equality of partition columns (table1.a = table2.a) -
+Shards of the distributed table should match one to one, i.e. each shard
+of table A should overlap with one and only one shard from table B.
+
+For example lets assume we 3 hash distributed tables X, Y and Z and let
+X and Y have 4 shards while Z has 8 shards.
+
+.. code:: sql
```
Comment (ozgune) [inline articles/outer_joins.rst]: Minor question.

I thought this blog post also had "circle images" that represented the type of joins. Did we take them out?

Comment (begriffs) [reply articles/outer_joins.rst]: Oops, added them.

**Comment by begriffs** [comment]
OK, I took care of your comments and noticed quite a few other things to improve as well.

> I wasn't able to see an active deployment / preview. How can we get a preview?

Heroku had taken the preview down due to inactivity. I've recreated it and it's at https://citus-doc-preview-pr-424.herokuapp.com/

> Since the content in the articles section is the same as the one in our blog, do we need to do anything on the SEO side?

Yes, after deployment to the docs I'll need to work with Jason or somebody else with CloudFlare access to set some Canonical URI headers for each of these pages to tell search engines that the blog posts are the real source. This will avoid SEO penalties, but it will also mean that google searches for terms in these articles will link to our blog rather than our docs. If we want the reverse behavior for search we will need to add those headers when serving the blog posts instead.

> We shortened or changed the names for these blog posts. Could we keep the original titles? (Similarly, one the of links was too short -- aggregation.rst. Could we make it a bit more descriptive?)

I shortened the names so they would fit better on the left hand nav menu. However I restored them all to their original long versions for you, and I guess it doesn't look so bad on the menu. Check out the preview and see what you think. Also a name like `aggregations.rst` only affects the URL. In the docs a link to this page would show "Real-time Event Aggregation at Scale Using Postgres with Citus," or else a custom title if that was specified in the originating link.

> We skipped over some of the articles we originally picked. For example, we skipped over the sharding for a multi-tenant app and HLL articles. Could we include them?

Done.

**Comment by ozgune** [comment]
Hey @begriffs -- I went over the preview and the changes look generally good.

I had the following comments before we could merge this pull request. **I'm super excited by this work btw.**

1. Could we start concurrent work on the SEO side (or this isn't feasible)? That way, we wouldn't penalize ourselves for too long?
2. In what order do we present these articles? Maybe we sort them by their order of publication on our blog?
3. Is there a way to signal to the reader that these Articles were previously published on our blog? Maybe we write a short intro for the Articles section? Or, include an "originally published" sentence at the top? (That way, the reader would know that some of these articles could be dated.)
4. The images in these articles have descriptions underneath them. Do we need those descriptions?
5. Could you click on all links before submitting to check that they are still live?
6. I have a few short comments on specific articles. I'll add them to the PR shortly.

File: articles/scale_on_aws.rst [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+process.
+
+The second and more important one relates to auto failover. If the
+primary node fails, how do you detect the failure and promote a
+secondary to be the new primary? (In distributed systems, this is also
+known as the leader election problem.) To enable this, you need a
+monitoring component that watches the primary and secondary nodes and
+reliably initiates their failovers.
+
+.. image:: ../images/citus-cloud-diagrams.png
```
Comment (ozgune) [inline articles/scale_on_aws.rst]: Small question. Is there a way to "center" the images in this article?

File: articles/scale_on_aws.rst [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+(pg\_stat views, table and index hit ratios), and active and long
+running queries.
+
+Once you decide on these metrics, you need to have a daemon that
+periodically collects these metrics, a solution to serialize and store
+these metrics, and a visualization layer to display them. A common stack
+that accomplishes this can run collectd on database instances, Graphite
+to store them, and Grafana to display them.
+
+.. image:: ../images/citus-cloud-monitoring.png
```
Comment (ozgune) [inline articles/scale_on_aws.rst]: This image looks too big. Could we make it smaller?

File: articles/outer_joins.rst [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+     FROM customer JOIN purchase ON customer.customer_id = purchase.customer_id
+     ORDER BY purchase.comment;
+
+          name      |        comment
+     ---------------+------------------------
+      Corra Ignacio | Comfortable
+      Jalda Francis | Good Read, cheap price
+      Corra Ignacio | Nice to Have!
+
+.. image:: ../images/articles-join-inner.png
```
Comment (ozgune) [inline articles/outer_joins.rst]: The images in this Article aren't aligned (as they were in the blog post).

This was a bit confusing to me. Is there a way to align them with the sql code blocks?

File: articles/outer_joins.rst [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+shard of table A should overlap with one and only one shard from table
+B.
+
+For example lets assume we 3 hash distributed tables X, Y and Z and let
+X and Y have 4 shards while Z has 8 shards.
+
+.. code:: sql
+
+    CREATE TABLE user (user_id int, name text);
+    SELECT master_create_distributed_table('user', 'user_id', 'hash');
```
Comment (ozgune) [inline articles/outer_joins.rst]: Could we swap these APIs (and the ones below) with `create_distributed_table`?

File: articles/semi_structured_data.rst [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+of queries such as:
+
+Top page views over the last 7 days for a given site:
+
+.. code:: sql
+
+    SELECT page, 
+           count(*)
+    FROM visits
+    WHERE site_id = ‘foo’
```
Comment (ozgune) [inline articles/semi_structured_data.rst]: The quotes in this code block (and the one below) look incorrect.

Could we change ‘foo’ with 'foo'?

File: articles/sharding_mt_app.rst [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+
+List the products for a particular store:
+
+.. code:: sql
+
+    SELECT id, 
+           name,
+           price
+    FROM products
+    WHERE store_id = foo;
```
Comment (ozgune) [inline articles/sharding_mt_app.rst]: The quoting around foo looks off.

Could we add quotes around foo here and change the quotes below to 'foo'?

File: articles/sharding_mt_app.rst [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+all tables. By doing this you can easily shard out all your data so it’s
+located on the same shard. In the above data model we coincidentally had
+``store_id`` on all of our tables, but if it weren’t there you could add
+it. This would put you in a good position to distribute all your data so
+it’s stored on the same nodes. So now lets try sharding our tenants, in
+this case stores:
+
+.. code:: sql
+
+    SELECT master_create_distributed_table('stores', 'id', 'hash');
```
Comment (ozgune) [inline articles/sharding_mt_app.rst]: Could we change these APIs to `create_distributed_table`?

File: articles/designing_saas.rst [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+much better way.
+
+.. figure:: ../images/articles-saas-fig-1.png
+   :alt: Sharding tables for multi-tenant applications
+
+   Sharding tables for multi-tenant applications
+
+Common benefits of having all tenants share the same database are:
+
+Resource pooling (reduced cost): If you create a separate database for
```
Comment (ozgune) [inline articles/designing_saas.rst]: We have a few words / sentences that are underlined in the original blog post. (I think "resource pooling", "if you shard ..." sentence, "ease of maintenance" and maybe 1-2 others).

Could we underline these phrases or sentences?

File: articles/designing_saas.rst [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+   Representing data that varies across tenants, modern method
+
+Of course, these aren’t the only design criteria and questions to be
+aware of. If you shard your database tables, how do you handle isolation
+or integrate with ORM libraries? What happens if you have a table that
+you can’t easily add a tenant\_id column? In this article, we focused
+on building multi-tenant databases with scaling as the primary
+consideration in mind; and skipped over certain points. If you’re
+looking to learn more about designing multi-tenant databases, please
+sign up for our `upcoming webinar on the
```
Comment (ozgune) [inline articles/designing_saas.rst]: Could we change this expired webinar link to say our slack channel?

**Comment by begriffs** [comment]
> Could we start concurrent work on the SEO side (or this isn't feasible)? That way, we wouldn't penalize ourselves for too long?

Putting the SEO changes into effect won't take very long. The main delay will be getting the time of someone who has CloudFlare access. I can put some time on Jason's calendar for this, and wait to merge this PR until that time. Also we are running low on the number of *page rules* that CloudFlare has allocated to us. We'll need more to add the headers for each of these articles. I'll ask Jason about his opinion here and we may need to pay more money to get more rules, or else find and replace old rules that aren't useful anymore.

> In what order do we present these articles? Maybe we sort them by their order of publication on our blog?

Kind of at random. Some of the articles are grouped by topic like the hyperloglog ones at the end. Not sure that sorting by publication date is all that relevant. In fact sorting in the opposite direction might be better, if anything, since older articles may not reflect the current state of Citus as well. :man_shrugging: 

> Is there a way to signal to the reader that these Articles were previously published on our blog? Maybe we write a short intro for the Articles section? Or, include an "originally published" sentence at the top? (That way, the reader would know that some of these articles could be dated.)

Sure. Probably an "originally published" link in each is best since people may or may not read an introductory section.

> The images in these articles have descriptions underneath them. Do we need those descriptions?

I see what you mean. That came from the program I'm using to translate the Markdown format to ReStructuredText. I'll get rid of the unnecessary captions.

> Could you click on all links before submitting to check that they are still live?

Good idea. I ran a web crawler on the articles and found two broken links. One of them is actually the awesome multi-tenant microsoft article from 2006! Microsoft took the article down in July. I located it on the wayback machine and turned it into a PDF and included it in our docs repo.

File: articles/scale_on_aws.rst (hunk 6 above)
Comment (begriffs) [reply articles/scale_on_aws.rst]: (Done)

File: articles/scale_on_aws.rst (hunk 7 above)
Comment (begriffs) [reply articles/scale_on_aws.rst]: (Done)

File: articles/outer_joins.rst (hunk 8 above)
Comment (begriffs) [reply articles/outer_joins.rst]: (Done)

**Comment by begriffs** [comment]
@ozgune I think all the issues you pointed out are now taken care of, except determining a particular order for the articles in the TOC.

**Comment by ozgune** [comment]
hey @begriffs, this looks good.

I'm adding my final four comments to this pull request.

1. When I visit the docs homepage, I see Articles as a separate extended section under our eight buttons: https://citus-doc-preview-pr-424.herokuapp.com/

This looks a bit off to me. Could we remove Articles links from the documentation home page? (Maybe, we make Articles one of these buttons in the future?)

2. "How Distributed Outer Joins on PostgreSQL with Citus Work" Article looks different than the original one in three ways:
 (a) Our outer join images are placed left and right in the Article. They appear consistently on the right hand side in the blog post.
 (b) The bullet point starting with "Outer joins should be between distributed(sharded) tables only" is cut out. (The original blog post also incorrectly puts these together into a paragraph.)
 (c) Minor. The image after "How Citus Processes OUTER JOINs When one-to-one matching between shards exists, then performing an outer join on large tables is equivalent to combining outer join results of corresponding shards." is right aligned in the article.

3. SEO changes

> Putting the SEO changes into effect won't take very long. The main delay will be getting the time of someone who has CloudFlare access. I can put some time on Jason's calendar for this, and wait to merge this PR until that time. Also we are running low on the number of page rules that CloudFlare has allocated to us. We'll need more to add the headers for each of these articles. I'll ask Jason about his opinion here and we may need to pay more money to get more rules, or else find and replace old rules that aren't useful anymore.

Sounds good. Could you put time on Jason's calendar? If we need to pay more to CloudFlare to make this work, that also sounds good.

4. Article ordering

> In what order do we present these articles? Maybe we sort them by their order of publication on our blog?

> Kind of at random. Some of the articles are grouped by topic like the hyperloglog ones at the end. Not sure that sorting by publication date is all that relevant. In fact sorting in the opposite direction might be better, if anything, since older articles may not reflect the current state of Citus as well. :man_shrugging:

Yes, I had "sorting in the opposite direction" in mind. That way, we'd get articles that reflect Citus' current state at the top. Could you change the Articles' order accordingly?

**Comment by begriffs** [comment]
OK, the SEO, article ordering, image alignment, and fixed frontpage are all ready now.

Assuming those were the last pieces of feedback, so time to merge.

**Pull request merged**